src/fluree/db/query/subject_crawl/rdf_type.cljc

```diff
@@ -0,0 +1,105 @@
+            [fluree.db.index :as index]
+            [fluree.db.dbproto :as dbproto]
+            [fluree.db.flake :as flake]
+            [fluree.db.util.core :as util :refer [try* catch*]]
+            [fluree.db.util.log :as log]
+            [fluree.db.util.schema :as schema-util]
+            [fluree.db.permissions-validate :as perm-validate]
+            [fluree.db.query.subject-crawl.common :refer [where-subj-xf result-af subj-perm-filter-fn filter-subject order-results]]))
+
+#?(:clj (set! *warn-on-reflection* true))
+
+
+(defn flakes-xf
+  [{:keys [db fuel-vol max-fuel error-ch vars filter-map permissioned?] :as _opts}]
+  (let [permissions (when permissioned?
+                      (subj-perm-filter-fn db))]
+    (fn [flakes port]
+      (go
+        (try*
+          ;; TODO: Right now we enforce permissions after the index-range call, but
+          ;; TODO: in some circumstances we can know user can see no subject flakes
+          ;; TODO: and if detected, could avoid index-range call entirely.
+          (let [flakes* (cond->> flakes
+                                 filter-map (filter-subject vars filter-map)
+                                 permissioned? permissions
+                                 permissioned? <?)]
+            (when (seq flakes*)
+              (async/put! port flakes*))
+
+            (async/close! port))
+          (catch* e (async/put! error-ch e) (async/close! port) nil))))))
+
+(defn subj-flakes-chan
+  "Returns a channel that has a stream of flakes grouped by subject id.
+  Always uses :spot index."
+  [{:keys [conn novelty t spot] :as db} error-ch vars {:keys [o] :as _where-clause}]
+  (let [rdf-type    (or (:value o)
+                        (get vars (:variable o)))
+        cid         (or (dbproto/-c-prop db :id rdf-type)
+                        (throw (ex-info (str "Invalid data type: " rdf-type)
+                                        {:status 400 :error :db/invalid-query})))
+        fflake      (flake/->Flake (flake/max-subject-id cid) -1 nil nil nil util/min-integer)
+        lflake      (flake/->Flake (flake/min-subject-id cid) util/max-integer nil nil nil util/max-integer)
+        cmp         (:comparator spot)
+        range-set   (flake/sorted-set-by cmp fflake lflake)
+        in-range?   (fn [node]
+                      (query-range/intersects-range? node range-set))
+        query-xf    (where-subj-xf {:start-test  >=
+                                    :start-flake fflake
+                                    :end-test    <=
+                                    :end-flake   lflake
+                                    :return-type :flake-by-sid})
+        resolver    (index/->CachedTRangeResolver conn (:spot novelty) t t (:async-cache conn))
+        tree-chan   (index/tree-chan resolver spot in-range? query-range/resolved-leaf? 1 query-xf error-ch)
+        return-chan (async/chan 10 (partition-by flake/s))]
+    (async/go-loop []
```

Comment:
I think this go-loop can be:

```clojure
(async/pipeline 10 return-chan cat tree-chan)
```

Reply:
I'm not sure... here I'm doing a buffer of 10 vs a pipeline would be parallel processing of 10. I'm going to leave as is for now but we can discuss if there is a better way. The buffer is intentional to slightly 'read ahead' of the data needed downstream as it should be fairly cheap to do so and then ultimately pre-load the next index before it is needed.